rqt_gui_py/src/rqt_gui_py/rclpy_spinner.py

```diff
@@ -23,17 +27,56 @@ def __init__(self, node):
         super().__init__()
         self._node = node
         self._abort = False
+        self._listener_queue = Queue()
 
     def run(self):
         qDebug('Start called on RclpySpinner, spinning ros2 node')
         executor = MultiThreadedExecutor()
         executor.add_node(self._node)
         while rclpy.ok() and not self._abort:
             executor.spin_once(timeout_sec=1.0)
+
+            # Remove subscribers and clients that have been registered for destruction
+            while not self._listener_queue.empty():
+                listener = self._listener_queue.get()
+                self.__destroy_listener__(listener)
+
         if not self._abort:
             qWarning('rclpy.shutdown() was called before QThread.quit()')
 
     def quit(self):  # noqa: A003
         qDebug('Quit called on RclpySpinner')
         self._abort = True
         super().quit()
+
+    def register_listeners_for_destruction(self, *listeners):
+        """
+        Add subscriptions or service clients to the spinner's queue for destruction.
+
+        Subscription and clients cannot be destroyed while the executor is spinning, these will be
+        added to the queue and destroyed after spin_once has completed.
+
+        :param *listeners: The listeners that are to be destroyed
+        """
+        for listener in listeners:
+            if not isinstance(listener, Subscription) and not isinstance(listener, Client):
+                qWarning('Invalid object of type "{}" called for listener destruction'.format(
+                         type(listener)))
+                continue
+            qDebug('Registering destruction for listener of type "{}"'.format(type(listener)))
+            self._listener_queue.put(listener)
+
+    def __destroy_listener__(self, listener):
```

Comment:
Why this style? why not `_destroy_listener`?

Reply:
Because external functions should absolutely not call this.

Reply:
> Prepending a single underscore (_) has some support for protecting module variables and functions (not included with from module import *). While prepending a double underscore (__ aka “dunder”) to an instance variable or method effectively makes the variable or method private to its class (using name mangling) we discourage its use as it impacts readability and testability and isn’t really private.

The trailing double underscores could go